diff --git a/libgupnp/gupnp-service-proxy.c b/libgupnp/gupnp-service-proxy.c
index 2b862bc..2532919 100644
--- a/libgupnp/gupnp-service-proxy.c
+++ b/libgupnp/gupnp-service-proxy.c
@@ -483,7 +483,9 @@ gupnp_service_proxy_send_action_hash_gi (GUPnPServiceProxy *proxy,
                                          GHashTable        *out_hash,
                                          GError           **error)
 {
-    return gupnp_service_proxy_send_action_hash (proxy, action, error, in_hash, out_hash);
+    GHashTable *out_hash_fake = g_hash_table_new(NULL, NULL);
+
+    return gupnp_service_proxy_send_action_hash (proxy, action, error, in_hash, out_hash_fake);
 }
 
 
